Frameworks/CoreGraphics/CGImage.mm

```diff
@@ -21,981 +21,660 @@
-            TraceWarning(TAG, L"Warning: colorSpace = NULL, assuming Gray based on bytesPerRow.");
-            colorSpace = CGColorSpaceCreateDeviceGray();
-        }
-
-        colorSpaceAllocated = true;
-    }
-
-    __CGSurfaceFormat format = _CGImageGetFormat(bitsPerComponent, bitsPerPixel, colorSpace, bitmapInfo);
-
-    if (format != _ColorIndexed) {
-        __CGSurfaceInfo surfaceInfo = __CGSurfaceInfo(((__CGColorSpace*)colorSpace)->colorSpaceModel,
-                                                      bitmapInfo,
-                                                      bitsPerComponent,
-                                                      bitsPerPixel >> 3,
-                                                      width,
-                                                      height,
-                                                      bytesPerRow,
-                                                      data,
-                                                      format);
+CGImageRef CGImageCreateWithPNGDataProvider(CGDataProviderRef source,
+                                            const CGFloat decode[],
+                                            bool shouldInterpolate,
+                                            CGColorRenderingIntent intent) {
+  RETURN_NULL_IF((source == nullptr) ||
+                 ![(NSObject *)source isKindOfClass:[NSData class]]);
 
-        newImage = new CGBitmapImage(surfaceInfo);
-    } else {
-        __CGSurfaceInfo surfaceInfo = _CGSurfaceInfoInit(width, height, _ColorBGR);
-
-        newImage = new CGBitmapImage(surfaceInfo);
-        void* pData = newImage->Backing()->LockImageData();
-        int stride = newImage->Backing()->BytesPerRow();
-
-        char* curPosOut = (char*)pData;
-        char* curPosIn = (char*)data;
-        for (unsigned y = 0; y < height; y++) {
-            char* curLineOut = curPosOut;
-            char* curLineIn = curPosIn;
-            for (unsigned x = 0; x < width; x++) {
-                BYTE c = *curLineIn;
-
-                if (c <= ((__CGColorSpace*)colorSpace)->lastColor) {
-                    BYTE* palette = (BYTE*)&((__CGColorSpace*)colorSpace)->palette[c * 3];
-                    *(curLineOut++) = *(palette++);
-                    *(curLineOut++) = *(palette++);
-                    *(curLineOut++) = *(palette++);
-                } else {
-                    *(curLineOut++) = 0;
-                    *(curLineOut++) = 0;
-                    *(curLineOut++) = 0;
-                }
-                curLineIn++;
-            }
-
-            curPosOut += stride;
-            curPosIn += width;
-        }
-
-        newImage->Backing()->ReleaseImageData();
-    }
+  NSData *sourceData = (__bridge NSData *)source;
+  CGImageRef imageRef =
+      _CGImageLoadPNG((void *)[sourceData bytes], [sourceData length]);
 
-    newImage->_provider = dataProvider;
+  RETURN_NULL_IF(!imageRef);
+  imageRef->SetInterpolate(shouldInterpolate).SetRenderingIntent(intent);
```

Comment:
clang format;

---
In reply to: [84143254](https://github.com/Microsoft/WinObjC/pull/1185#discussion_r84143254) [](ancestors = 84143254)